Fix reg.py for use with LGPO module

twangboy opened the pull request:
### What does this PR do?
Modifies the `salt.utils.to_unicode` function to return `None` if passed `None`

Makes sure the `vdata` value type matches what is required by the type specified in `vtype`. The LGPO module was passing a unicode string 0 (`u'0'`) which was causing the RegSetValueEx function to fail because it was expecting an int type value.

This casts vdata to the following:
int for REG_DWORD and REG_QWORD
str for REG_SZ and REG_EXPAND_SZ
bin for REG_BINARY
list for REG_MULTI_SZ

Fixes many Unicode registry issues.
Uses `salt.utils.to_unicode` instead of the local `_unicode_to_mcbs` and `_mcbs_to_unicode` functions
Was using a mixture of _winreg and win32api... moved to win32api for better unicode handling

Tested with Unicode values
Tested with Unicode value names

Fixes `reg.py` state to handle Unicode value names and values
Can't pass Unicode values in the `name` portion of a state for some reason

### What issues does this PR fix or reference?
https://github.com/saltstack/salt/issues/44192

### Previous Behavior
LGPO would fail

### New Behavior
Now it works

### Tests written?
No

twangboy committed:
Cast vdata to it's proper type

twangboy committed:
Check type before casting

twangboy committed:
Use six.text_type instead of str

twangboy committed:
Fix many issues with reg.py

Remove _winreg dependency
Do ALL the things with win32api
Use salt.utils.to_unicode for all unicode conversions
Only use mbcs for returned values
salt.utils.to_unicode returns None if the passed value is None

twangboy committed:
make salt.utils.to_unicode return none when passed none

damon-atkins commented on salt/modules/reg.py:
```diff
@@ -69,104 +69,76 @@ def __virtual__():
-        return obj
-    else:
-        return _mbcs_to_unicode(obj)
+    return salt.utils.to_unicode(vdata, 'utf-8')
```
Why is this 'utf-8' instead of 'mbcs'

twangboy replied:
'mbcs' wasn't translating unicode characters correctly in my test environment. We're going straight utf-8 into the pywin32 functions and sending any output through the `_to_mbcs` function. From what I was able to gather, `mbcs` just means the default encoding for the system rather than any specific type of encoding.

damon-atkins replied:
mbcs is the encode for the language in use. The assumption is the data in the registry is this.  So if it was a Japanese version of windows mbcs..   Its also doco some where that this is what winreg uses to encode.    PY3 fixes all this mess.    It also maybe dependant on the version of pywin.

damon-atkins replied:
Their are some tests for reg and unicode.

twangboy replied:
The data in the registry is unicode or ascii (https://msdn.microsoft.com/en-us/library/windows/desktop/ms724884(v=vs.85).aspx). The functions we're using to put stuff in the registry accept unicode as input. We're using 'utf-8' because it seems to handle most cases and works well. We could try utf-16, but I didn't do any testing with it.

We're converting the return data from those functions (`reg.read_value`) to whatever the locale encoding is. That's handled by the `_to_mbcd` function.

damon-atkins replied:
Use regedit to place this in the registry  发送时间   I think it means date
In python 2 you will get ???? in python 3 you will get \u53d1\u9001\u65f6\u95f4

cachedout replied:
This seems like a valid concern. @twangboy were you able to look at this?

twangboy replied:
This has always been an issue in Py2.

damon-atkins commented on salt/modules/reg.py:
```diff
@@ -39,10 +39,10 @@
 
 # Import third party libs
 try:
-    from salt.ext.six.moves import winreg as _winreg  # pylint: disable=import-error,no-name-in-module
```
I see changes removing winreg/_winreg are you moving away from winreg?

twangboy replied:
Yes

damon-atkins commented on salt/modules/reg.py:
```diff
@@ -555,42 +490,49 @@ def set_value(hive,
         vtype=REG_LIST vdata='[a,b,c]'
     '''
-
-    if PY2:
```
From a performance point of view test IF something is PY2  vs an IF for every string inside _to_unicode() i.e. one IF vs many IFs   When PY2 is dropped it makes it easy to pull the old code out.  As the conversion to Unicode will no loner be needed.

twangboy replied:
The conversion will be removed in `salt.utils.to_unicode` when we go away from Py2

damon-atkins replied:
I would have thought that the function salt.utils.to_unicode from a PY3 point of view would convert bytes to unicode.    salt.utils.to_unicode  this does not need to be called in PY3 at all inside reg.py.  Why then call a function which does 0 but has an overhead..  I would have thought dropping PY2 support for windows is less  18 months away because PY2 is not up to the job. (might be wishful thinking)

twangboy replied:
It makes the code easier to maintain.

twangboy committed:
Fix various issues

Fix state to handle unicode
Simplify type conversion logic
Handle binary values correctly

rallytime commented:
@twangboy Lint: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/15933/violations/ 

:)

twangboy committed:
Remove unused import (lint)

damon-atkins commented:
See https://github.com/pywin32/pywin32/blob/master/win32/src/PyUnicode.cpp

```
#if (PY_VERSION_HEX < 0x03000000)
  return PyString_FromStringAndSize(str, len);
#else
  return PyUnicode_DecodeMBCS(str, len, "ignore"); or it  will be
  return PyUnicode_DecodeMBCS(str, len, "NULL);
#endif

```

twangboy commented:
Rebased...
